target/linux/ath79/patches-6.6/102-gpio-mmio-do-not-calculate-bgpio_bits-via-ngpios.patch

```diff
@@ -0,0 +1,33 @@
+From: Shiji Yang <yangshiji66@outlook.com>
+Date: Mon, 24 Jun 2024 03:56:08 +0000
+Subject: [PATCH] gpio: mmio: do not calculate bgpio_bits via "ngpios"
+
+bgpio_bits must be aligned with the data bus width. For example, on a
+32 bit big endian system and we only have 16 GPIOs. If we only assume
+bgpio_bits=16 we can never control the GPIO because the base address
+is the lowest address.
+
+low address                          high address
+-------------------------------------------------
+|   byte3   |   byte2   |   byte1   |   byte0   |
+-------------------------------------------------
+|    NaN    |    NaN    |  gpio8-15 |  gpio0-7  |
+-------------------------------------------------
+
+Fixes: 55b2395e4e92 ("gpio: mmio: handle "ngpios" properly in bgpio_init()")
+Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
+---
+ drivers/gpio/gpio-mmio.c | 2 --
+ 1 file changed, 2 deletions(-)
+
+--- a/drivers/gpio/gpio-mmio.c
++++ b/drivers/gpio/gpio-mmio.c
+@@ -622,8 +622,6 @@ int bgpio_init(struct gpio_chip *gc, str
+ 	ret = gpiochip_get_ngpios(gc, dev);
+ 	if (ret)
+ 		gc->ngpio = gc->bgpio_bits;
+-	else
```

Comment:
This does fix the bug, but I’d rather see it in a form that’s appropriate for upstreaming. Upstream did make this change intentionally, I don’t think they’d take this patch as-is. It would be nice to not have to maintain another local hack patch forever, and there is still a real bug lurking in the upstream.

I think that my `umax` in my version might be closer to acceptable (https://github.com/markmentovai/openwrt/commits/ath79_gpio-mmio_bgpio_bits), but still a little “magical”. Really, I think that a sentinel `sz` (like 0 or -1) to trigger calculating `bgpio_bits` from `ngpio` would be best.

One more thing: since I cracked this in https://github.com/openwrt/openwrt/issues/15739#issuecomment-2184600307 but you then took it upon yourself to write a new patch, I think that some form of credit (like a `Suggested-By:`) would be appropriate. And since Loránt was so diligent in testing, I think it’s also right to credit him with `Tested-By`.

Reply:
> I think that my `umax` in my version might be closer to acceptable (https://github.com/markmentovai/openwrt/commits/ath79_gpio-mmio_bgpio_bits), but still a little “magical”.

`umax` won't change the value because the max size of the `ngpio` is the bus width. It's hardware limit.

> Really, I think that a sentinel `sz` (like 0 or -1) to trigger calculating `bgpio_bits` from `ngpio` would be best.

Good idea. Maybe you can try work a patch and upstream it.

> One more thing: since I cracked this in [#15739 (comment)](https://github.com/openwrt/openwrt/issues/15739#issuecomment-2184600307) but you then took it upon yourself to write a new patch, I think that some form of credit (like a `Suggested-By:`) would be appropriate. And since Loránd was so diligent in testing, I think it’s also right to credit him with `Tested-By`.

And @lorand-horvath If you like. Please provide your name and email.

Reply:
@DragonBluep Name: Lóránd Horváth and email: lorand.horvath82@gmail.com Thanks!

Reply:
Looking at [the commit that introduced that line](https://github.com/torvalds/linux/commit/55b2395e4e92adc492c6b30ac109eb78250dcd9d), I think the change is fine, and using `bgpio_bits` less than `sz * 8` cannot work for big endian, and is dubious at best for little endian. What e.g. if you need to use 32 bit reads/writes for accessing the register? So I think it has a decent chance of being accepted upstream, if argued well enough.

Going by the intention of the original commit, rounding up (`sz * 8`) to the next power of two would fix it in case one passes a `sz` that is not a power of two. Or reject `sz` values that are not a power of two. Or just leave it as is, as `bgpio_setup_accessors()` will just complain and fail if it isn't, and maybe we shouldn't try to fix up user's mistakes, and instead reject broken configurations, so they can be fixed. Assuming there is even anyone doing that.

So yeah, I think just removing the round up line is fine, and shouldn't add any regressions.

Reply:
I read the thrust of [the upstream change](https://github.com/torvalds/linux/commit/55b2395e4e92adc492c6b30ac109eb78250dcd9d) as being that `ngpio` can imply `bgpio_bits`, but in the case `ngpio` isn’t available, `bgpio_bits` can imply `ngpio`. In most cases, it should be possible to properly configure the driver by specifying only one of the two.

That falls apart for platforms like ath79, which has an architectural reason for `bgpio_bits` to be 32 to accommodate `ngpio` up to 24 in later chips, but ar71xx only has 12 `ngpio`. So that’s why we’re here.

But I read the upstream commit as saying that it shouldn’t be strictly necessary to supply both. It’s saying that it’s reasonable to tell the driver “here are 32 GPIOs, you figure it out”. But I think it’s just as reasonable to need to tell the driver, as we do in this case, “here are 12 GPIOs, but the registers are 32 bits wide”.

Just because ath79 needs this special accommodation doesn’t mean that it should be imposed on other GPIO devices generally. That’s why I’m saying that the upstream change was made for a reason (but see below), and I think that throwing away the `else` clause might be going against something that was intentional. The behavior was specifically called out in the commit message, so I don’t think that it was an accident.

Anyway, my whole point was that we shouldn’t do something that’s specific to ath79 and that can’t go upstream. If the upstream will take this, that’s great. I can certainly argue the other side: you need to know both the register width and the number of GPIOs. My thinking was just that upstream seemed to already accept the argument that you don’t, so it might be an uphill battle.

I dug a little further and found that the `else` we’re talking about was originally introduced between [v3](https://lore.kernel.org/lkml/202303050354.HH9DhJsr-lkp@intel.com/t/) and [v4](https://lore.kernel.org/lkml/3e99caed-13bb-204c-4458-e0670fa9e8c9@gmail.com/t/), based on [this part of the discussion](https://lore.kernel.org/lkml/202303050354.HH9DhJsr-lkp@intel.com/t/#m546051fb98777e40d29a0576b7502b749ee93d48). Honestly I do see the logic as shaky there, but it seems to have been accepted. I think that what Andy was really asking for there was just a comment saying affirmatively that `bgpio_bits = sz * 8` is correct.

Reply:
> Anyway, my whole point was that we shouldn’t do something that’s specific to ath79 and that can’t go upstream. 

I'm pretty sure any big endian users where `ngpio <= (sz * 8)/2` is broken by this, ath79 was just the one where we noticed it.

> I can certainly argue the other side: you need to know both the register width and the number of GPIOs.

Right, that's why `bgpio_init()` has the `sz` argument, which is the register size. Therefore `bgpio_bits` cannot be anything else that `sz * 8`, regardless how many bits are actually used for GPIOs.

If we do support the following one

> But I read the upstream commit as saying that it shouldn’t be strictly necessary to supply both. It’s saying that it’s reasonable to tell the driver “here are 32 GPIOs, you figure it out”.

Then `bgpio_bits` should only be calculated from `ngpios` if `sz` passed was `0`.

So if we would want to support something like this, we would want this

```
if (!gc->bgpio_bits)
   gc->bgpio_bits = roundup_pow_of_two(round_up(gc->ngpio, 8));
```

but then we need to re-do the check for bgpio_bits being too large.

Reply:
https://github.com/openwrt/openwrt/pull/15784#discussion_r1651424744
> Then `bgpio_bits` should only be calculated from `ngpios` if `sz` passed was `0`.

Yup, that’s what I suggested at the beginning (https://github.com/openwrt/openwrt/pull/15784#discussion_r1651152196):
> Really, I think that a sentinel `sz` (like 0 or -1) to trigger calculating `bgpio_bits` from `ngpio` would be best.

In your judgment, do you think that would be most palatable upstream? Or do you still want to propose just dropping the `else`?

Reply:
> [#15784 (comment)](https://github.com/openwrt/openwrt/pull/15784#discussion_r1651424744)
> 
> > Then `bgpio_bits` should only be calculated from `ngpios` if `sz` passed was `0`.
> 
> Yup, that’s what I suggested at the beginning ([#15784 (comment)](https://github.com/openwrt/openwrt/pull/15784#discussion_r1651152196)):
> 
> > Really, I think that a sentinel `sz` (like 0 or -1) to trigger calculating `bgpio_bits` from `ngpio` would be best.
> 
> In your judgment, do you think that would be most palatable upstream? Or do you still want to propose just dropping the `else`?

I would probably just drop it instead of trying to guess what the original author intended, and then go from there based on the comments that gets.

Arguing with that this can cause the bits to get reduced if `ngpio` is less than half of `sz * 8`, which then reduces the accessor width, which then breaks big endian systems since now only the upper byte(s) or the registers get accessed.

Reply:
Anyway, I've sent the patch upstream. Let's see if they have a better solution. I'll update this PR later.
https://lore.kernel.org/linux-gpio/TYCP286MB089577B47D70F0AB25ABA6F5BCD52@TYCP286MB0895.JPNP286.PROD.OUTLOOK.COM/

Reply:
Thank you, Shiji!

Reply:
@DragonBluep Do you have any news from upstream? When is this supposed to be accepted/rejected?

Reply:
> @DragonBluep Do you have any news from upstream? When is this supposed to be accepted/rejected?

You can follow the progress of the current [submission on the linux-gpio patchwork](https://patchwork.ozlabs.org/project/linux-gpio/patch/TYCP286MB089577B47D70F0AB25ABA6F5BCD52@TYCP286MB0895.JPNP286.PROD.OUTLOOK.COM/).

I tried to address all comments, so it's currently up to the maintainers to pick it up/comment on it.